### PR TITLE
[Api] orders index for customer's account

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | part of #11250 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/DataProvider/OrderCollectionDataProvider.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+use ApiPlatform\Core\DataProvider\CollectionDataProviderInterface;
+use ApiPlatform\Core\DataProvider\RestrictedDataProviderInterface;
+use Sylius\Bundle\ApiBundle\Context\UserContextInterface;
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/DataProvider/OrderCollectionDataProvider.php:24]:
```suggestion

/** @experimental */
```

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/OrderCollectionDataProviderSpec.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        OrderInterface $secondOrder
+    ): void {
+        $userContext->getUser()->willReturn($user);
+        $user->getRoles()->willReturn(['ROLE_API_ACCESS']);
+        $orderRepository->findAll()->willReturn([$firstOrder, $secondOrder]);
+
+        $this->getCollection(OrderInterface::class)->shouldReturn([$firstOrder, $secondOrder]);
+    }
+
+    function it_provides_orders_created_by_logged_uses(
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/OrderCollectionDataProviderSpec.php:52]:
```suggestion
    function it_provides_orders_created_by_logged_user(
```

**Review by Tomanhez — APPROVED** [review]
And rebase needed

File: features/account/customer_account/viewing_orders_history/seeing_customer_orders_placed_as_guest.feature [hunk 3]
```diff
@@ -11,19 +11,11 @@ Feature: Seeing customer's orders placed as guest
...
-        And I decide to change my address
-        And I specify the email as "ned@stark.com"
-        And I complete the addressing step
-        And I select "Free" shipping method
-        And I complete the shipping step
-        And I select "Offline" payment method
-        And I complete the payment step
-        And I confirm my order
-        And I register with previously used "ned@stark.com" email and "lannistersAreDumb" password
+        Given the customer "ned@stark.com" created account with password "lannistersAreDumb"
```
Comment (GSadee) [inline features/account/customer_account/viewing_orders_history/seeing_customer_orders_placed_as_guest.feature:18]:
```suggestion
        Given I register with previously used "ned@stark.com" email and "lannistersAreDumb" password
```

File: src/Sylius/Behat/Context/Api/Shop/AccountContext.php [hunk 4]
```diff
@@ -189,6 +205,48 @@ public function iShouldBeNotifiedThatElementIsInvalid(): void
...
+     */
+    public function iBrowseMyOrders(): void
+    {
+        $this->orderShopClient->index();
+    }
+
+    /**
+     * @When I register with previously used :email email and :password password
+     */
+    public function iRegisterWithPreviouslyUsedEmailAndPassword(string $email, string $password): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/AccountContext.php:219]: Couldn't it be extracted to `RegistrationContext`?

File: src/Sylius/Behat/Context/Api/Shop/AccountContext.php [hunk 5]
```diff
@@ -189,6 +205,48 @@ public function iShouldBeNotifiedThatElementIsInvalid(): void
...
+     */
+    public function iShouldSeeASingleOrderInTheList(): void
+    {
+        Assert::same($this->responseChecker->countCollectionItems($this->orderShopClient->index()), 1);
+    }
+
+    /**
+     * @Then this order should have :order number
+     */
+    public function thisOrderShouldHaveNumber(OrderInterface $order): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/AccountContext.php:240]:
```suggestion
    public function thisOrderShouldHaveNumber(string $orderNumber): void
```
?

File: src/Sylius/Behat/Context/Setup/UserContext.php [hunk 6]
```diff
@@ -61,6 +61,20 @@ public function thereIsUserIdentifiedBy($email, $password = 'sylius')
         $this->userRepository->add($user);
     }
 
+    /**
+     * @Given the customer :customerEmail created account with password :password
+     */
+    public function theCustomerCreatedAccountWithPassword($email, $password = 'sylius')
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Setup/UserContext.php:67]:
```suggestion
    public function theCustomerCreatedAccountWithPassword(string $email, string $password = 'sylius'): void
```

File: src/Sylius/Bundle/ApiBundle/DataProvider/OrderCollectionDataProvider.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        $this->orderRepository = $orderRepository;
+    }
+
+    public function getCollection(string $resourceClass, string $operationName = null)
+    {
+        /** @var ShopUserInterface $user */
+        $user = $this->userContext->getUser();
+
+        if ($user instanceof AdminUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return $this->orderRepository->findAll();
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/DataProvider/OrderCollectionDataProvider.php:45]: I guess that we shouldn't provide carts for admin users

File: src/Sylius/Bundle/ApiBundle/DataProvider/OrderCollectionDataProvider.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        if ($user instanceof ShopUserInterface) {
+            /** @var CustomerInterface $customer */
+            $customer = $user->getCustomer();
+
+            return $this->orderRepository->findByCustomer($customer);
+        }
+
+        return null;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/DataProvider/OrderCollectionDataProvider.php:55]:
```suggestion
        return [];
```
?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 9]
```diff
@@ -42,6 +42,16 @@
                     <attribute name="summary">Pickups a new cart</attribute>
                 </attribute>
             </collectionOperation>
+
+            <collectionOperation name="shop_get">
+                <attribute name="method">GET</attribute>
+                <attribute name="path">/shop/orders</attribute>
+                <attribute name="normalization_context">
+                    <attribute name="groups">
+                        <attribute>order:shop:read</attribute>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:51]: I would be for:
```suggestion
                        <attribute>shop:order:read</attribute>
```

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/OrderCollectionDataProviderSpec.php (hunk 2 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/OrderCollectionDataProviderSpec.php:52]:
```suggestion
    function it_provides_orders_created_by_logged_in_user(
```

File: src/Sylius/Behat/Context/Api/Shop/AccountContext.php (hunk 4 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Context/Api/Shop/AccountContext.php:219]: could be... but if I do that I get many names conflicts, so I left it here.

File: src/Sylius/Bundle/OrderBundle/Doctrine/ORM/OrderRepository.php [hunk 10]
```diff
@@ -125,4 +125,14 @@ public function findCartsNotModifiedSince(\DateTimeInterface $terminalDate): arr
             ->getResult()
         ;
     }
+
+    public function findAllExceptCarts(): array
```
Comment (GSadee) [inline src/Sylius/Bundle/OrderBundle/Doctrine/ORM/OrderRepository.php:129]: I'm not convinced to that name, but for now, I have no better idea

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Adam! :1st_place_medal:

**Pull request merged**